docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,186 @@
+is a problem in practice. UniFFI came from a world where we had Rust crates and
+a hand-written FFI that exposed types from all over the crate. If these tools
+had started with the limitations from the macro approach in mind, it's possible
+a different, acceptable design might have been made to work. Maybe duplicating
+some structs and supplying suitable `Into` implementations might make things workable?
+
+Diplomat comes from a very smart team. They may well come up with a novel solution, so
+UniFFI should track the progress of that project to see what we can gleefully steal
+in the future. As discussed below, a kind of "hybrid" approach might even be possible.
+
+# Looking forward
+
+Before looking forward, let's step back a little - both UniFFI and diplomat are solving the exact
+same use-cases, just using a different approach to defining the type universe.
+But if we ignore that, the tools take the same basic approach - they all build the
+type universe, then use the representation of this type universe to define both Rust
+and foreign bindings.
+
+The type universe described by diplomat is somewhat "leaner" than that described by UniFFI -
+Rust types are the first-class citizens in the universe. UniFFI defines an external type model -
+for example, there's a `Type` enum where, for example, `Type::Record(Record)` represents a
+Rust struct. In other words, diplomat's type world can not be divorced from Rust,
+whereas UniFFI's already is.
+
+That said though, there might be a future where merging or otherwise creating some
+interoperability between these type universes might make sense. You could imagine
+a world where you can use diplomat to describe your type universe, but use UniFFI's foreign
+generation code to generate the Kotlin bindings. Similarly, a world where you use UniFFI
+and UDL files to describe your type universe, but then use diplomat to generate
+the NodeJS bindings.
+
+Or to put it another way, you could imagine a world where both tools are split into a
+"describe the type universe" portion and a "build the bindings" portion, and these tools
+could be used together.
+
+Sadly, that looks like alot of work, so someone would probably need to find a compelling
+actual use-case to perform this work.
+
+# Next steps for UniFFI
+
+As much as some of the UniFFI team dislike the external UDL file, there's no clear path to
+moving away from it. The macro approach is too limiting, and no other promising opportunities
+have presented themselves. There's no clear alternative to UDL which allows a complex
+type universe to be described, and at this stage, any replacement would need to be
+compelling enough to make a change worthwhile, which is hard to imagine.
+
+In the short term, the best we can probably do is to enumerate the perceived problems
+with the UDL file and try to make them more ergonomic - for example, avoiding repetition of
+`[Throws=SomeError]` would remove alot of noise, and some strategy for generating
```

Comment:
diplomat just treats documentation as "yet another backend", which works reasonably well, since the architecture of a diplomat backend is just "here's the type structure, you know what to expect on the FFI layer, do what you want".